[service-bus] Session manager sample preview.1

richardpark-msft opened the pull request:
Built a sample that users can use as a basis for building their own round-robin session receiver.

As part of this I made two changes in the core package:
1. Made it so creating a session receiver is now an async API. This simplified the sample quite a bit and was part of the work we were going to do for preview.2 anyways.
2. Made it so the "I asked for next available session but got nothing" says so in the message rather than the old error which complained about mismatched session IDs.

richardpark-msft commented:
Note to reviewers - our goal is to determine if this is "simple" enough to be written as a sample. I believe I've handled most of the stuff that we generally try to offer - namely:
- auto lock renewal
- sessions are read to completion
- using push so the user can get things like autocomplete

Do we think the sample is too complex? Do we want to code-golf it a little?

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+  processMessage(msg: ReceivedMessageWithLock);
+  processError(err: Error, sessionId: string);
+  processClose(reason: "error" | "idle_timeout", sessionId: string);
+}
```
If we had to make the sample simpler, I would remove this interface.

Assuming that the intention of this sample is to just show the building blocks rather than provide an implementation that user can copy/paste and call into, all 4 handlers above can be removed.

- Remove the `processInitialize` and `processClose` handlers. We can replace calls to these with a comment saying that this is the place the user can do set-up/tear-down for the session
- Remove the `processMessage and `processError` handlers. The console.log() we have now can be inlined.

I would also make the assumption that the idle timeout feature is built-into the session receiver i.e pass the timeout value in options to `createSessionReceiver()`

richardpark-msft replied:
The balance here is tough.

There's much more code in this sample than in a normal sample and there's multiple spots where the user would need to fill stuff in. We could remove this but it does make it a bit easier to follow, IMO. 

My vote is still to keep them - I actually didn't have them originally and added them in after trying to just read through the sample and see if it was "easy" to figure out which parts to fill in.

bterlson replied:
I worry people will expect they need to write this interface. Can we just make process* bare functions and call them from the appropriate places? It might be a bit more verbose but simpler usage.

richardpark-msft replied:
@ramya-rao-a , @bterlson  - I think what's in there now should make both of you happy :)

- The functions are at file scope now so we still get some nice defined area for the users to see what they need to fill in.
- No more interface. BAD BAD INTERFACE.

ramya-rao-a replied:
I would rename `processInitialize` to `processSessionInitialize` and `processClose` to `processSessionClose`. Otherwise as a user I am 20 lines into the sample and now seeing something that I am not sure why I would need to use.

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+  processClose(reason: "error" | "idle_timeout", sessionId: string);
+}
+
+function _createIdleTimer(
```
It would be nice if this was supported out of the box by the session receiver. But I can see that the challenge is going to be communicating to the caller that we stopped receiving messages because of the timeout vs any kind of error.

richardpark-msft replied:
Maybe we could just export it as a primitve, similar to what we do for `delay`?

richardpark-msft replied:
I made it a bit shorter which should hopefully ease the pain :)

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+  try {
+    await sessionReceiver.close();
+  } catch (err) {
+    await handlers.processError(err, sessionReceiver.sessionId);
```
I wouldnt put a try/catch around close() in a sample without first being clear on the kind of errors we would expect from it and the expected reactions from the customer.

richardpark-msft replied:
This is a good point. I believe we do have a larger conversation scheduled to talk about this so I don't mind having it here and then just adding doc comments to `close()` when we do finally do that.

ramya-rao-a replied:
I do mind having it here and in no other samples.
If needed, you can log an issue to have the larger conversation and at the end of it update the samples as needed.

richardpark-msft replied:
Removed.

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+      },
+      async processError(err) {
+        threwError = true;
+        await handlers.processError(err, sessionReceiver.sessionId);
```
All errors other than the ones stemming from running `processMessage` should be considered as fatal and should result in us moving to the next session.

richardpark-msft replied:
I need to check - do we funnel errors thrown by the user in processMessage here too? Do we provide an 'isFatal' flag or something on our internal errors?

ramya-rao-a replied:
Yes, we do funnel the errors thrown by the user to `processError`.
From what I recall, an error can be an instance of the custom class `MessagingError` (in which case there is a retryable property on it) or it could be on of the common error class like `Error` and `TypeError`.

So, all retryable `MessagingError` errors should be treated as non fatal and allow moving to the next session

ramya-rao-a replied:
But, on the other hand `SessionLockLostError` is not retryable, but I would like to keep the loop going for this error...

richardpark-msft replied:
I simplified it some more to make it so that an error or the idle timeout cause us to close the session receiver rather than try to dive into error handling here.

richardpark-msft replied:
CC: @ramya-rao-a

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+
+  // if we get a fatal error in any of the promises we can just shut it all down
+  console.log(`All session promises have been started`);
+  await Promise.race(allPromises);
```
What is the reason to wait for the promise returned by `_roundRobinThroughAvailableSessions` to resolve?
Looks like if there was no fatal error from individual sessions, then this promise would never resolve.

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+    }
+  };
+
+  await _roundRobinThroughAvailableSessions(handlers, abortController.signal);
```
What is the reason to wait for the promise returned by `_roundRobinThroughAvailableSessions` to resolve?
Looks like if there was no fatal error from individual sessions, then this promise would never resolve.

richardpark-msft replied:
The .push() and all of that is gone. We just let the promises float free.

ramya-rao-a commented:
> Do we think the sample is too complex? 

Its definitely not simple :) Can we have a pull counterpart to this to see how it fares in terms of complexity?

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+
+  for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
+    allPromises.push(
+      new Promise(async (_res, rej) => {
```
We can skip creating the new promise and the try/catch here if we move the while loop into `_processNextSession` right? Or create a new function 

```javascript
async function keepProcessingSessions(serviceBusClient, handlers, abortSignal) {
      while (!abortSignal.aborted) {
            await _processNextSession(serviceBusClient, handlers, abortSignal);
          }
}
```

Simplifies the for loop to 

```js
for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
   allPromises.push(keepProcessingSessions(serviceBusClient, handlers, abortSignal))
}

richardpark-msft replied:
I still kept the while loop on the outside, but I did eliminate the Promise (and also the push, which removed even more).

I like having the while loop out of the function so the function doesn't have this extra level of indentation, etc... if people want to pick through it.

richardpark-msft commented:
> > Do we think the sample is too complex?
> 
> Its definitely not simple :) Can we have a pull counterpart to this to see how it fares in terms of complexity?

I think the pull sample will just be more complex (also, I sent out that preview link where I did have that in it). It'll seem less complex until you realize you have to complete/abandon messages, etc... (which is our general argument in favor of using `subscribe` vs `receiveBatch` anyways).

With regard to the overall complexity - the subject matter itself is complex too but we're going to have to explain it in some fashion. Code might be the simplest way. We also have some prerequisites before they even try this:
- They're in the `advanced` samples folder. This isn't beginner level material.
- They already understand sessions.
- They've probably already used receivers.

Also, while this *is* more complex than our usual SDK samples but I think it's not actually as complex as samples we might offer in other avenues, like when you can download a whole-app style sample from docs.microsoft.com.

Aside from the feedback you provided in your PR comments was anything else nagging at the back of your mind about it?

bterlson commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,155 @@
+  };
+}
+
+async function _noSessionsAvailable(err: Error) {
```
I'm not sure the leading underscore is warranted here

richardpark-msft replied:
I just wanted some way to say "this is a private detail".

I ended up eliminate this particular function but I'm guessing you want it gone from the others as well? Is there a better convention?

richardpark-msft replied:
Removed for now. :)

richardpark-msft commented:
@ramya-rao-a , @chradek , @HarshaNalluru  - I've added in some more code. I mostly wanted to make sure I was able to get what I wanted working and it all seemed. There are some TODOs but those will have to wait for next week.

So far:
- `async createSessionReceiver()` is a thing and is in there. As a result of that I was able to guarantee that sessionId was filled out so that went from being a `string | undefined` to a `string`
- I throw a specific error if I detect that we sent in undefined as the session ID (ie, get next) and we received undefined as the session ID. I'm 100% sure that I can't just fabricate new `MessagingError.code` values though so I'll need some guidance there.

TODOs:
- There's still some work that Ramya mentioned that I need to get to for the session receiver sample
- I'd like to get the .code stuff correct - there's clearly a methodology to our errors.

Please review at your nearest convenience. :)

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts:
```diff
@@ -0,0 +1,132 @@
+  for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
+    (async () => {
+      while (!abortSignal.aborted) {
+        await _processNextSession(serviceBusClient, abortSignal);
```
If `_processNextSession` throws an error, what happens to other concurrent sessions? Would they keep going or should we abort them?

richardpark-msft replied:
There was one spot where we propagated an error and actually threw (when we would createSessionReceiver). I changed this to be consistent and just funnel the error through the user's processError() function instead.

I think we're okay with the simple "just keep going forever" model. I moved the abortController definition up a bit higher and mentioned (in a comment) that you can use `abortController.abort` to terminate the session manager.

ramya-rao-a commented:
Overall feedback:
- Need to be clear on what errors should result in gracefully close the current session and move on to the next session. In sample, this would result in figuring out how to control closing the session from inside the `processError` callback.
- Need to be clear on whether we want to advocate use of promises that never resolve, but only reject

richardpark-msft commented:
I've addressed these two I believe:

> Overall feedback:
> 
> * Need to be clear on what errors should result in gracefully close the current session and move on to the next session. In sample, this would result in figuring out how to control closing the session from inside the `processError` callback.

I changed it so the sample just treats all errors as fatal. Given that this is a sample having some simple error handling seems like enough. If the errors are temporary in nature we'll loop back around and get the session again if it wasn't empty.

> * Need to be clear on whether we want to advocate use of promises that never resolve, but only reject

I removed it so the promises really just stay alive forever and never resolve or reject. That might not have been the direction you were thinking :)

Let me know if you have some feedback on the sample, or the way that I exposed the error when we find that there are no more available sessions.

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,146 @@
+
+  await processInitialize(sessionReceiver.sessionId);
+
+  const sessionFullyRead = new Promise((resolveSessionAsFullyRead, rejectSessionWithError) => {
```
When the timer runs out, the promise `sessionFullyRead` is resolved.
When there is an error, the promise `sessionFullRead` is regjected

But, neither ensures that we stop receiving messages right away.
In this sample, we await on `processClose()` and then close the session receiver.

Am guessing this is because we currently don't have a way to stop receiving messages without closing the message receiver? If so, then until we have this feature, can we pass an abort signal to `subscribe()` instead?

richardpark-msft replied:
We do pass an abortSignal to subscribe?

ramya-rao-a replied:
I meant that the subscribe method is not stopped right away after the promise is resolved/rejected
If the subscribe method returned a closable thing, then we would call it before calling `processClose`.

The abort signal being passed is the uber signal that is passed to the main function of this sample. I was referring to an abort signal that would get fired for when `sessionFullyRead` is resolved or rejected. You can use link the two signals such that subscribe is aborted when either is fired

richardpark-msft replied:
And in both cases the receiver is probably about to exit w/o any more messages _anyways_:
* The session is probably empty and isn't getting any more messages (the reason the idle timer fired in the first place).
* An error happened in which case rhea is about to close the link anyways.

We could add more error handling code in here but for sample purposes I think it's covered well enough.

ramya-rao-a replied:
Fair point on the second point, but there is a corner case of users `processClose()` taking long enough for a message to come by in case of idle timeout :)

But I agree, for the sake of this sample, we are good.

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,146 @@
+  } catch (err) {
+    if ((err as MessagingError).code === "SessionCannotBeLockedError") {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
```
I see this delay in the same vein as the delay we have in between retries. If so, then why is it that only the `SessionCannotBeLockedError` case gets to benefit from it? Why not all the errors?

richardpark-msft replied:
It's possible they do - this one is the only one I definitely knew I should delay for.

If there are others that make sense that wouldn't be covered by the normal retry logic let me know.

ramya-rao-a replied:
There is no normal retry logic atleast at the moment. We don't recover broken receiver links when it comes to sessions. So in my eyes, all errors would need this delay. If so, then the only difference is whether you want to call `processError` for `SessionCannotBeLockedError` or not.

For the sake of simplicity, I would treat all the caught errors as the same i.e call `processError`, do the delay and return.

richardpark-msft replied:
Sounds like that's the best option then.

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,146 @@
+
+  for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
+    (async () => {
+      while (!abortSignal.aborted) {
```
When the abort signal is fired, we stop spinning any more sessions, but what about already created session receivers, how would they stop?

richardpark-msft replied:
subscribe() is also passed the same abortSignal.

ramya-rao-a replied:
Ah! subscribe is getting the signal directly from the controller. I would either expect the signal to be passed to `receiveFromNextSession()` or have `roundRobinThroughAvailableSessions()` access it directly from the controller too.

richardpark-msft replied:
Ah, I meant to use the global abortController just to simplify things.

Fixed (by removing the parameters altogether)

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,146 @@
+  This sample demonstrates how you can continually read through all the available
+  sessions in a Service Bus queue or subscription.
+
+  Run the sendMessages sample with some session ids before running this sample.
```
```suggestion
  Run the sendMessages sample with different session ids before running this sample.
```

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,146 @@
+// this can be used control when the round-robin processing will terminate.
+const abortController = new AbortController();
+
+// called just before we start processing the first message of a session
```
> called just before ...

I worry that this phrasing gives the impression that it gets passed to the service library and that the library will call it. Especially because `processMessage` and `processError` which do get passed to the library have the same phrasing. Can we rephrase that this is something that the user needs to call?

>  .. just before we start processing the first message of a session

Should we make it clear that this is called before we start receiving at all? I wonder if the current text makes it feel that we receive the first message, but before we pass it to `processMessage`, we call the `processInitialize`

My recommendation:

> Helper function for you to call once a session receiver is initialized, but before receiving any messages

And then for `processMessage`

> Helper function to be passed to the subscribe() method on the receiver to be called for each received message

richardpark-msft replied:
I'm not sure I understand the concern. The user can easily see how these functions are used.

Also, the rewording doesn't seem accurate either - the user isn't expected to call these for the sample at all - that's all handled for them.

ramya-rao-a replied:
> that's all handled for them

handled by the sample vs handled by the library: that is the distinction I want to make

`processMessage` and `processError` are called by the library
`processInitialize` and `processClose` are called by the sample. And the sample for all means and purposes is user code, not library code

richardpark-msft replied:
My point is - what problem are you trying to solve here? When we had an interface I could see the issue but the functions and their call points are clearly documented and viewable.

What confusion do you think is going to happen on the user's side? That they'll go to the main service package and try to use these methods there?

richardpark-msft replied:
After offline discussion with @ramya-rao-a we agreed to modify the wording. I've also changed the names of the functions to better distinguish them from the actual handler methods that are passed into `subscribe`.

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,146 @@
+
+// called if we get an error
+async function processError(err: Error, sessionId?: string) {
+  console.log(`[${sessionId}] had error ${err.message}`);
```
Given that we are marking `sessionId` as optional here, I would add an if/else here.
I would also print the entire error as MessagingError has the code which is important
```js
if (sessionId) {
    console.log(`Error when receiving messages from the session ${sessionId}: `, err);
} else {
    console.log(`Error when creating the receiver for next available session`, err);
}
```

richardpark-msft replied:
Done!

ramya-rao-a commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,146 @@
+    console.log(`[${sessionId}] was closed because of error`);
+  } else if (reason === "idle_timeout") {
+    console.log(`[${sessionId}] was closed - no more messages within idle timeout`);
+  }
```
Since you have captured the two reasons in the comments above the function here, we can skip the same in code. Or skip the comment and keep the code

```js
console.log(`[${sessionId}] was closed because of ${reason}`)
```

richardpark-msft replied:
Done!

ramya-rao-a commented on sdk/servicebus/service-bus/src/serviceBusClient.ts:
```diff
@@ -227,8 +227,7 @@ export class ServiceBusClient {
 
-    // TODO: .NET actually tries to open the session here so we'd need to be async for that.
-    return new SessionReceiverImpl(
+    const sessionReceiverImpl = SessionReceiverImpl.createInitializedSessionReceiver(
```
I would recommend that we either add an await here or directly return.
One would expect that a const named `sessionReceiverImpl` to have the actual implementation, but here it has the promise

richardpark-msft replied:
Ah, meant to remove that. :)

ramya-rao-a commented on sdk/servicebus/service-bus/src/receivers/sessionReceiver.ts:
```diff
@@ -141,6 +140,43 @@ export class SessionReceiverImpl<ReceivedMessageT extends ReceivedMessage | Rece
+        "Internal error: open() should not be called after the SessionReceiver has been created"
+      );
+    }
+    return this._createMessageSessionIfDoesntExist();
```
Can we not skip the check on the message session and the following error altogether if we move `return this._createMessageSessionIfDoesntExist()` into `createInitializedSessionReceiver()`

ramya-rao-a replied:
Ah! its because `createInitializedSessionReceiver` is static and `_createMessageSessionIfDoesntExist` is not. Then my follow up would be, do we need the static method at all? Why not create the SessionReceiverImpl as before and then call init() ?

richardpark-msft replied:
I can eliminate the check. I did this just so I didn't have to make `_createMessageSessionIfDoesntExist` public. 

Now that I've locked in the constructor+init() pairing I think I'll just bit the bullet and do that now.

richardpark-msft replied:
Done.

ramya-rao-a commented on sdk/servicebus/service-bus/src/receivers/sessionReceiver.ts:
```diff
@@ -141,6 +140,43 @@ export class SessionReceiverImpl<ReceivedMessageT extends ReceivedMessage | Rece
+    // `createInitializedSessionReceiver` will set this value by calling init()
+    // so we just temporarily set it to "" so we can get away with it never being
+    // `undefined`.
+    this.sessionId = "";
```
This makes me slightly nervous... Turns out empty string is a valid session id...

richardpark-msft replied:
Yes, it's part of the reason why I made it so it's not possible to construct a SessionReceiver directly. The window of time here is between constructor() and init() which is small and manageable.

HarshaNalluru commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,144 @@
+// `reason` will be:
+// * 'error' if we are closing because of an error(the error will be delivered
+//   to `processError` above)
+// * 'idle_timeout' if we `sessionIdleTimeoutMs` milliseconds pass without
```
```suggestion
// * 'idle_timeout' if `sessionIdleTimeoutMs` milliseconds pass without
```
?

HarshaNalluru commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,144 @@
+} from "@azure/service-bus";
+import * as dotenv from "dotenv";
+import { env } from "process";
+import { AbortController } from "@azure/abort-controller";
```
Add @azure/abort-controller in the package.json for the samples ?

richardpark-msft replied:
Oh man, yes. I've been debugging this using the main project. Good catch!

richardpark-msft replied:
Fixed!

HarshaNalluru commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,144 @@
+    });
+  } catch (err) {
+    if ((err as MessagingError).code === "SessionCannotBeLockedError") {
+      console.log(`INFO: no available sessions, sleeping for ${delayOnErrorMs}`);
```
`delayOnErrorMs` is not defined

richardpark-msft replied:
That was weird. I've been staging this in hunks and I wonder if I pushed something incomplete. It is there now (if you view all commits).

HarshaNalluru commented on sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts:
```diff
@@ -0,0 +1,147 @@
+const sessionIdleTimeoutMs = 3 * 1000;
+const delayOnErrorMs = 5 * 1000;
+
+// this can be used control when the round-robin processing will terminate.
+const abortController = new AbortController();
```
Maybe show an example?

richardpark-msft replied:
Done! I actually had an example closer to the actual call to start the session round robining as well.

richardpark-msft commented:
> 
> 
> Are we adding the JS version of the sample in a separate PR?

Nope, generating this now.

richardpark-msft commented:
/azp run js - servicebus - tests

richardpark-msft commented:
/azp run js - servicebus - tests

richardpark-msft commented:
/azp run js - servicebus - tests
